command/agent/agent.go

```diff
@@ -172,82 +191,206 @@ func Create(config *Config, logOutput io.Writer, logWriter *logger.LogWriter, re
+		srv, err := NewDNSServer(a, &c.DNSConfig, a.LogOutput, c.Domain, a.dnsAddr.String(), c.DNSRecursors)
+		if err != nil {
+			return fmt.Errorf("error starting DNS server: %s", err)
+		}
+		a.dnsServers = []*DNSServer{srv}
+	}
+
+	// start HTTP servers
+	return a.startHTTP(a.httpAddrs)
+}
+
+func (a *Agent) startHTTP(httpAddrs map[string][]net.Addr) error {
+	// ln contains the list of pending listeners until the
+	// actual server is created and the listeners are used.
+	var ln []net.Listener
+
+	// cleanup the listeners on error. ln should be empty on success.
+	defer func() {
+		for _, l := range ln {
+			l.Close()
+		}
+	}()
+
+	// bind to the listeners for all addresses and protocols
+	// before we start the servers so that we can fail early
+	// if we can't bind to one of the addresses.
+	for proto, addrs := range httpAddrs {
+		for _, addr := range addrs {
+			switch addr.(type) {
+			case *net.UnixAddr:
+				switch proto {
+				case "http":
+					if _, err := os.Stat(addr.String()); !os.IsNotExist(err) {
+						a.logger.Printf("[WARN] agent: Replacing socket %q", addr.String())
+					}
+					l, err := ListenUnix(addr.String(), a.config.UnixSockets)
+					if err != nil {
+						return err
+					}
+					ln = append(ln, l)
+
+				default:
+					return fmt.Errorf("invalid protocol: %q", proto)
+				}
+
+			case *net.TCPAddr:
+				switch proto {
+				case "http":
+					l, err := ListenTCP(addr.String())
+					if err != nil {
+						return err
+					}
+					ln = append(ln, l)
+
+				case "https":
+					tlscfg, err := a.config.IncomingTLSConfig()
+					if err != nil {
+						return fmt.Errorf("invalid TLS configuration: %s", err)
+					}
+					l, err := ListenTLS(addr.String(), tlscfg)
+					if err != nil {
+						return err
+					}
+					ln = append(ln, l)
+
+				default:
+					return fmt.Errorf("invalid protocol: %q", proto)
+				}
+
+			default:
+				return fmt.Errorf("invalid address type: %T", addr)
+			}
+		}
 	}
 
-	return agent, nil
+	// https://github.com/golang/go/issues/20239
+	//
+	// In go1.8.1 there is a race between Serve and Shutdown. If
+	// Shutdown is called before the Serve go routine was scheduled then
+	// the Serve go routine never returns. This deadlocks the agent
+	// shutdown for some tests since it will wait forever.
+	//
+	// We solve this with another WaitGroup which checks that the Serve
+	// go routine was called and after that it should be safe to call
+	// Shutdown on that server.
+	var up sync.WaitGroup
+	for _, l := range ln {
+		l := l // capture loop var
+
+		// create a server per listener instead of a single
+		// server with multiple listeners to take advantage
+		// of the Addr field for logging. Since the server
+		// does not keep state and they all share the same
+		// agent there is no overhead.
+		addr := l.Addr().String()
+		srv := NewHTTPServer(addr, a)
+		a.httpServers = append(a.httpServers, srv)
+
+		up.Add(1)
+		a.wgServers.Add(1)
+		go func() {
+			defer a.wgServers.Done()
+			up.Done()
```

Comment:
It does not guarantee it. Until the Go team fixes the bug I think this is the best we can do with the minimal amount of overhead. This guarantees that the go routine that executes `Serve` gets at least scheduled once before `Start` returns. My understanding is that it yields again on the next blocking call which should happen within `Serve`. I don't see how it would yield between `up.Done()` and `Serve` until I got it wrong.